halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
     }
 }
 
+/// Data that needs to be preprocessed from a circuit
+#[derive(Debug, Clone)]
+pub struct PreprocessingV2<F: Field> {
+    // TODO(Edu): Can we replace this by a simpler structure?
+    pub(crate) permutation: permutation::keygen::Assembly,
```

Comment:
Happy to have a call and explore if we can simplify this!!!

Reply:
I've started planning the crate split and had to address this.  I chose to keep this information as a vector pairs of cells, where each cell is a column + row_offset.  Then the backend receives this and loads it into the `permutation::keygen::Assembly` type.